library/src/scala/Tuple.scala

```diff
@@ -248,36 +304,67 @@ object Tuple {
   def unapply(x: EmptyTuple): true = true
 
   /** Convert an array into a tuple of unknown arity and types */
-  def fromArray[T](xs: Array[T]): Tuple = {
+  def fromArray[T](xs: Array[T]): Tuple =
+    fromArray(xs, xs.length)
+
+  /** Convert the first `n` elements of an array into a tuple of unknown arity and types */
+  def fromArray[T](xs: Array[T], n: Int): Tuple = {
     val xs2 = xs match {
       case xs: Array[Object] => xs
       case xs => xs.map(_.asInstanceOf[Object])
     }
-    runtime.Tuples.fromArray(xs2)
+    runtime.Tuples.fromArray(xs2, n)
   }
 
   /** Convert an immutable array into a tuple of unknown arity and types */
-  def fromIArray[T](xs: IArray[T]): Tuple = {
+  def fromIArray[T](xs: IArray[T]): Tuple = fromIArray(xs, xs.length)
+
+  /** Convert the first `n` elements of an immutable array into a tuple of unknown arity and types */
+  def fromIArray[T](xs: IArray[T], n: Int): Tuple = {
     val xs2: IArray[Object] = xs match {
       case xs: IArray[Object] @unchecked => xs
       case _ =>
         xs.map(_.asInstanceOf[Object])
     }
-    runtime.Tuples.fromIArray(xs2)
+    runtime.Tuples.fromIArray(xs2, n)
   }
 
   /** Convert a Product into a tuple of unknown arity and types */
   def fromProduct(product: Product): Tuple =
     runtime.Tuples.fromProduct(product)
 
+  extension [X <: Tuple](inline x: X)
+
+    /** The index (starting at 0) of the first occurrence of y.type in the type `X` of `x`
+     *  or Size[X] if no such element exists.
+     */
+    transparent inline def indexOf(y: Any): Int = constValue[IndexOf[X, y.type]]
```

Comment:
I think this is misleading to only look at the type of the argument, typically this would be some runtime operation based on equals, why not make it explicitly an operation on types?
```suggestion
    transparent inline def indexOf[T]: Int = constValue[IndexOf[X, T]]
```

Reply:
It _was_ an operation on types before Eugene's comment, then we changed it. We can have both `indexofOfType` and `indexOf`.

Reply:
It was on operation on a type inferred via a term parameter
```scala
transparent inline def indexOf[T](y: T): Int = constValue[IndexOf[X, T]]
```
Jamie's proposed version requiring an _explicit_ type parameter looks even better to me.

Reply:
That's _exactly_ what we had before [164005a](https://github.com/scala/scala3/pull/19174/commits/164005a669cc9c0bb7f5201e401d5ca1ce5a3a2a)

    /** The index (starting at 0) of the first element in the type `X` of `x`
     *  that matches type `Y`.
     */
    inline def indexOfType[Y] = constValue[IndexOf[X, Y]]

Reply:
Ah my bad, sorry about that. 
I guess I got tripped up by the comment that used to be below
```
/* Note: It would be nice to add the following two extension methods:
      inline def indexOf[Y: Precise](y: Y) = constValue[IndexOf[X, Y]]
      inline def containsType[Y: Precise](y: Y) = constValue[Contains[X, Y]]
    because we could then move indexOf/contains completely to the value level.
    But this requires `Y` to be inferred precisely, and therefore a mechanism
    like the `Precise` context bound used above, which does not yet exist.
    */
```